src/Illuminate/Container/BoundMethod.php

```diff
@@ -139,40 +158,73 @@ protected static function getCallReflector($callback)
+    protected static function addDependenciesToInputData(array $signature, array $inputData)
     {
-        if (array_key_exists($parameter->name, $parameters)) {
-            $dependencies[] = $parameters[$parameter->name];
-
-            unset($parameters[$parameter->name]);
-        } elseif ($parameter->getClass() && array_key_exists($parameter->getClass()->name, $parameters)) {
-            $dependencies[] = $parameters[$parameter->getClass()->name];
-
-            unset($parameters[$parameter->getClass()->name]);
-        } elseif ($parameter->getClass()) {
-            $dependencies[] = $container->make($parameter->getClass()->name);
-        } elseif ($parameter->isDefaultValueAvailable()) {
-            $dependencies[] = $parameter->getDefaultValue();
+        // Here we iterate through the list of declared parameters (in the method signature) and decide
+        // whether it should be invoked with the provided input data, or we should resolve an object
+        // for it (according to it's type-hint) or just call it with it's defined "default" value.
+        $resolvedInputData = [];
+        $i = 0;
+
+        foreach ($signature as $parameter) {
+            if (array_key_exists($parameter->name, $inputData)) {
+                $resolvedInputData[] = $inputData[$parameter->name];
+            } elseif ($class = $parameter->getClass()) {
+                $resolvedInputData[] = self::getInstance($inputData, $class->name, $i);
+            } elseif (array_key_exists($i, $inputData)) {
+                $resolvedInputData[] = $inputData[$i++];
+            } elseif ($parameter->isDefaultValueAvailable()) {
+                $resolvedInputData[] = $parameter->getDefaultValue();
+            }
         }
+
+        return $resolvedInputData;
     }
 
     /**
      * Determine if the given string is in Class@method syntax.
      *
-     * @param  mixed  $callback
+     * @param  string|callable  $callback
      * @return bool
      */
     protected static function isCallableWithAtSign($callback)
     {
         return is_string($callback) && strpos($callback, '@') !== false;
     }
+
+    /**
+     * @param array $inputData
+     * @param $className
+     * @param $i
+     * @return mixed
+     */
+    protected static function getInstance(array $inputData, $className, &$i)
+    {
+        if (array_key_exists($className, $inputData)) {
+            // gets from associative array input data
+            return $inputData[$className];
+        } elseif (isset($inputData[$i]) && is_a($inputData[$i], $className)) {
+            // gets from indexed array input data
+            return $inputData[$i++];
+        } else {
+            // resolves a new instance
+            return self::$container->make($className);
+        }
+    }
+
+    /**
+     * @param ContainerContract $container
+     */
+    public static function setContainer(ContainerContract $container)
```

Comment:
This should be nullable as well as you are passing null here in some cases (the same applies to the `call` method).